### ISSUE TITLE
Assertion failed: transformSourceNode cannot use DAG, must have single trail.

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/fourier-making-waves#73]
To reproduce:

1. Go to any of the 3 screens
2. Tab until you get to the navigation bar.
3. When tab from the Discrete navbar icon to the Wave Game navbar icon, this Error occurs:

```
Assertion failed: transformSourceNode cannot use DAG, must have single trail.
assert.js:25 Uncaught Error: Assertion failed: transformSourceNode cannot use DAG, must have single trail.
    at window.assertions.assertFunction (assert.js:25)
    at FocusHighlightFromNode.getUniqueHighlightTrail (FocusHighlightPath.js:238)
    at HighlightOverlay.activateHighlight (HighlightOverlay.js:237)
    at HighlightOverlay.onFocusChange (HighlightOverlay.js:507)
    at TinyProperty.emit (TinyEmitter.js:86)
    at Property._notifyListeners (Property.js:271)
    at Property.set (Property.js:186)
    at Property.set value [as value] (Property.js:341)
    at Function.set focus [as focus] (Display.js:2174)
    at Object.focus (PDOMPointer.js:59)
```

The problem appears to be with the icon created by `createHomeScreenIcon` in DiscreteScreen.js.  As far as I can tell, I'm not using any DAG features here, so I've been unable to figure out what is causing this.   The error message is also not particularly helpful.

@jessegreenberg any ideas? And/or anything you can tell me about PDOM limitations on scenery DAG features?

**Comment by pixelzoom** [comment]
Note that this icon contains some actual sliders, instantiated in `createHomeScreenIcon`:

```js
102 children: _.map( harmonics, harmonic => new AmplitudeSlider( harmonic, emphasizedHarmonics ) )
```

But ScreenIcon sets its iconNode to `pickable:false`, so my expectation is that these sliders will not affect the default pDomOrder of this icon.

**Comment by pixelzoom** [comment]
I modified FocusHighlightPath to identify "transformSourceNode":

```js
  getUniqueHighlightTrail() {
    assert && assert( this.transformSourceNode.instances.length <= 1,
      `transformSourceNode cannot use DAG, must have single trail, transformSourceNode=${this.transformSourceNode}` );
    return this.transformSourceNode.getUniqueTrail();
  }
```

... and the error message is:

```
Assertion failed: transformSourceNode cannot use DAG, must have single trail, transformSourceNode=GrippyThumb#15
```

GrippyThumb is the custom thumb for AmplitudeSlider.  No idea why it's causing this error, or why it's apparently being traversed.   Is something that is `pickable: false` still traversed?  How do I tell PDOM **NOT** to traverse its `iconNode` argument?

**Comment by pixelzoom** [comment]
I tried explicitly setting `pDomOrder` on the `iconNode` passed to ScreenIcon:

```js
  iconNode.pDomOrder = [];
// or
  iconNode.pDomOrder = null;
```

... but that changed nothing.

**Comment by pixelzoom** [comment]
Wild guess, I tried creating my sliders with `focusable: false`:

```js
    children: _.map( harmonics, harmonic => new AmplitudeSlider( harmonic, emphasizedHarmonics, {
      focusable: false
    } ) )
```

... and that fixes the problem.

I also tried setting `focusable: false` on the entire iconNode:

```js
  const iconNode = new Node( {
    children: [ xAxis, sliders ],
    focusable: false
  } );
```

... and that does NOT fix the problem.  Apparently settting a parent Node to `focusable: false` does not prevent its children from getting focus, not at all what I expected. 

Two questions:

(1) I frequently create icons programmatically, so this is a rather general problem.  Do I really need to set each interactive UI component used in an icon to `focusable: false`

(2) Why does this result in an assertion failure related to DAG? As far as I can tell, this has nothing to do with DAG, I'm not reusing any Nodes, and not using any scenery DAG features.

**Comment by jessegreenberg** [comment]
I worked on this today, and https://github.com/phetsims/scenery/issues/1235 was the result. 

I confirmed that DAG is being used in this case by joist for the ScreenIcon and HomeScreenButton:
<details>

![image](https://user-images.githubusercontent.com/6396244/121445075-bd00ff00-c95e-11eb-9639-94a82ca67b11.png)
</details>

There is a feature of focus highlights where you can create a focus highlight around one Node, track the transform of that Node so it also transforms highlight, and set that highlight as a highlight for different Node. This is used in Slider, where the focus highlight for the Slider actually surrounds the thumb
```js
this.focusHighlight = new FocusHighlightFromNode( thumb );
```

This feature doesn't work generally with DAG since to track the transform we need a unique Trail. In https://github.com/phetsims/scenery/issues/1235 I added another search to find the right Trail that should work almost all of the time. And when it doesn't work, there is now a clear assertion message that describes the issue and ways to avoid it.

For this issue, if you remove https://github.com/phetsims/fourier-making-waves/commit/0f6aca035b9d9ce0618c7407b634ebe5c0313530 and tab through the home screen or navigation bar, focus moves through the sliders in the icon now without error. But that isn't the behavior we want. I don't think we will ever want the icons within the HomeScreenButton and NavigationBarScreenButton to be focusable/readable so I made the icons `pdomVisible: false` in those types and removed the workaround in DiscreteScreen.

@pixelzoom assigning back to you to review and verify this is fixed, if you wish.

**Comment by pixelzoom** [comment]
Thanks @jessegreenberg.  The joist fix looks good, and thanks for removing the Fourier workaround.  And thanks for the nice explanation of what was going on - I now understand the "odd" behavior, and it makes sense to me.

We're likely to run into this in other places. For example, any checkbox or button might have "content" that was programmatically created from focusable UI components.   What do you think we should do to address those cases? Should we make the client responsible for setting `pdomVisible: false` on the content? Or should that be the responsibility of the UI components?

**Comment by jessegreenberg** [comment]
I am leaning toward making that the responsibility of the UI components were possible. A checkbox icon shouldn't be focusable and I don't think it should ever have content that should be read by a screen reader.  I made issues in sun and scenery-phet to review components and make icons `pdomVisible: false` where appropriate so this kind of won't happen for components in those repos.

Anything else for this issue?

**Comment by pixelzoom** [comment]
> I am leaning toward making that the responsibility of the UI components were possible. 

That sounds OK to me. I've already run into this in Fourier, for a checkbox icon, and `pdomVisible: false` worked fine.

Closing.
